src/core/app.lua

```diff
@@ -124,6 +128,65 @@ function configure (new_config)
    counter.add(configs)
 end
 
+-- Claims a name for a program so it can be identified by name by other processes.
+--
+-- The name given to the function must be unique, if a name has been used before by
+-- an active process the function will error displaying an appropriate error message.
+-- Any other process can enumerate programs by their name and PID using enumerate_named_programs.
+function claim_name(name)
+   configuration[name] = name
+   local namedir = "by-name/" .. name
+   local namedirqualified = named_program_root .. "/" .. name
+   local piddir = shm.root .. "/" .. S.getpid()
+
+   -- Verify that the by-name directory exists.
+   shm.mkdir(namedir)
+   
+   -- Check if a symlink to a named program with the same name exists, if not, unlink it.
+   if S.lstat(namedirqualified) then
+      -- It exists, lets check what the PID is.
+      local piddir = S.readlink(namedirqualified)
+      local ps, pe = string.find(piddir, "/[^/]*$")
+      local pid = tonumber(string.sub(piddir, ps+1, pe))
+      if S.kill(pid, 0) then
+         -- It's a running process, we should display an error message.
+         error("Name has been claimed by a currently running process: "..namedirqualified)
```

Comment:
Okay so.

I've understood the issue with stale PIDs. There is some mitigation against any problems they would cause.

1. `claim_name` checks for stale PID's using kill 0 and if it is dead, it'll unlink it.
2. `enumerate_named_programs` also uses kill 0 on each named program it discovers and only enumerates those which are still alive.
3. I've added to the shutdown function in which is run by the supervisor to unlink named programs which are being shutdown. 

This should mean that directories are cleaned up, it also should mean if something happens and they're not cleaned up programs can still claim the name as it will be found at that stage and unlinked so it can be re-used and finally other programs enumerating these would only get named programs which are still active at the time of calling `enumerate_named_programs`